File: java/VendingMachine/test/com/ravendarque/unitTests/RailShould.java
package com.ravendarque.unitTests;

import com.ravendarque.vendingMachine.rails.Rail;
import com.ravendarque.vendingMachine.rails.RailEmptyException;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class RailShould {

    private static final double DUMMY_PRICE_1 = 1;
    private static final int DUMMY_INVENTORY_0 = 0;
    private static final String DUMMY_LABEL = "Dummy label";

    private static final int TEST_INVENTORY_0 = 0;
    private static final int TEST_INVENTORY_1 = 1;

    @Test
    void returnTrueWhenCheckingIfRailWithNoItemsIsEmpty() {

        final Rail testRail = new Rail(DUMMY_INVENTORY_0, DUMMY_PRICE_1, DUMMY_LABEL);

        final boolean railIsEmpty = testRail.isEmpty();

        assertTrue(railIsEmpty);
    }

    @Test
    void returnFalseWhenCheckingIfRailWithOneItemIsEmpty() {

        final Rail testRail = new Rail(TEST_INVENTORY_1, DUMMY_PRICE_1, DUMMY_LABEL);

        final boolean railIsEmpty = testRail.isEmpty();

        assertFalse(railIsEmpty);
    }

    @Test
    void returnTrueWhenCheckingIfRailIsEmptyAfterLastItemHasBeenVended()
            throws RailEmptyException {

        final Rail testRail = new Rail(TEST_INVENTORY_1, DUMMY_PRICE_1, DUMMY_LABEL);
        testRail.vend();

        final boolean railIsEmpty = testRail.isEmpty();

        assertTrue(railIsEmpty);
    }

    @Test
    void provideAPrice() {

        final double expectedRailPrice = 1;

        final double testRailPrice = 1;
        final Rail testRail = new Rail(DUMMY_INVENTORY_0, testRailPrice, DUMMY_LABEL);

        final double actualRailPrice = testRail.getPrice();

        assertEquals(expectedRailPrice, actualRailPrice);
    }

    @Test
    void throwExceptionOnVendWhenRailIsEmpty() {

        final Rail rail = new Rail(TEST_INVENTORY_0, DUMMY_PRICE_1, DUMMY_LABEL);

        assertThrows(RailEmptyException.class, rail::vend);
    }

    @Test
    void provideALabel() {

        final String expectedLabel = "Test label";

        final String testLabel = "Test label";
        final Rail testRail = new Rail(DUMMY_INVENTORY_0, DUMMY_PRICE_1, testLabel);

        final String actualLabel = testRail.getLabel();

        assertEquals(expectedLabel, actualLabel);
    }

    @Test
    void throwExceptionWhenPriceIsNegative() {

        final double testPrice = -1;

        assertThrows(IllegalArgumentException.class, () -> new Rail(
                DUMMY_INVENTORY_0, testPrice, DUMMY_LABEL));
    }
}

File: C++/VendingMachine/VendingMachine.UnitTestsG/RailShould.cpp
#include "pch.h"
#include "../VendingMachine/Rail.h"
#include "../VendingMachine/RailEmptyError.h"

static const int DUMMY_CAPACITY_1 = 1;
static const int DUMMY_INITIAL_INVENTORY_0 = 0;
static const double DUMMY_PRICE_0 = 0;
static const std::string DUMMY_LABEL = "Dummy label";

TEST(RailShould, ReturnTrueWhenCheckingIfRailWithNoItemsIsEmpty)
{
	const Rail testRail(DUMMY_INITIAL_INVENTORY_0, DUMMY_PRICE_0, DUMMY_LABEL);

	EXPECT_TRUE(testRail.IsEmpty());
}

TEST(RailShould, ReturnFalseWhenCheckingIfRailWithOneItemIsEmpty)
{
	const auto testInitialInventory = 1;

	const Rail testRail(testInitialInventory, DUMMY_PRICE_0, DUMMY_LABEL);

	EXPECT_FALSE(testRail.IsEmpty());
}

TEST(RailShould, ProvideAPrice)
{
	const auto expectedPrice = 1;

	const auto testPrice = 1;

	const Rail testRail(DUMMY_INITIAL_INVENTORY_0, testPrice, DUMMY_LABEL);

	const auto actualPrice = testRail.GetPrice();

	EXPECT_EQ(actualPrice, expectedPrice);
}

TEST(RailShould, ProvideALabel)
{
	const auto expectedLabel = "Test label";

	const std::string testLabel = "Test label";

	const Rail testRail(DUMMY_INITIAL_INVENTORY_0, DUMMY_PRICE_0, testLabel);

	const auto actualLabel = testRail.GetLabel();

	EXPECT_EQ(actualLabel, expectedLabel);
}

TEST(RailShould, ReturnTrueWhenCheckingIfRailIsEmptyAfterLastItemHasBeenVended)
{
	const auto testInitialInventory = 1;

	Rail testRail(testInitialInventory, DUMMY_PRICE_0, DUMMY_LABEL);
	testRail.Vend();

	EXPECT_TRUE(testRail.IsEmpty());
}

TEST(RailShould, ThrowErrorOnVendWhenRailIsEmpty)
{
	Rail testRail(DUMMY_INITIAL_INVENTORY_0, DUMMY_PRICE_0, DUMMY_LABEL);

	EXPECT_THROW(testRail.Vend(), RailEmptyError);
}

TEST(RailShould, ThrowErrorWhenPriceIsNegative)
{
	const auto testPrice = -1;

	EXPECT_THROW(
		Rail testRail(DUMMY_INITIAL_INVENTORY_0, testPrice, DUMMY_LABEL);,
		std::invalid_argument);
}
